### PR TITLE
BUG: fix an infinite loop in itk::VoronoiDiagram2DGenerator

### PR DESCRIPTION
**Pull request by ILoveGoulash**
The loop exit condition was present in the code, but never wired and removed via https://github.com/InsightSoftwareConsortium/ITK/commit/cd97879e6bfec6cd9a33a3dd374d68cf4ff4303e

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by dzenanz** [comment]
@bradking Is ghostflow acting up? Commit [62457de](https://github.com/InsightSoftwareConsortium/ITK/commit/62457dec8703b4506501660fd1520abec4397b8a) has a very short commit message.

**Comment by ILoveGoulash** [comment]
Nah, the full "http://..." link is probably what triggers it, GitHub prettifies it on display.

File: Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx [hunk 1]
```diff
@@ -274,8 +274,10 @@ VoronoiDiagram2DGenerator<TCoordinate>::ConstructDiagram()
     buildEdges.push_back(curr);
     EdgeInfo front = curr;
     EdgeInfo back = curr;
-    while (!(rawEdges[i].empty()))
+    auto     maxStop = static_cast<int>(rawEdges[i].size());
```
Comment (hjmjohnson) [inline Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx]: Why is there a static_cast to a signed int here?  What if there are more edges than will fit in the positive range of an integer?

I think maxStop could just be the same type as size() ( likely size_t ).

Comment (ILoveGoulash) [reply Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx]: The `> 0` condition can't work on an unsigned. It'd make sense to cast it to a bigger type or use `size_t` and check for `!= SIZE_MAX`, though. Which one do you prefer?

Comment (N-Dekker) [reply Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx]: @ILoveGoulash Thank you for your pull request. 👍  Specifically regarding your comment here, I think you're mistaken! When `u` is an unsigned number, the condition `u > 0` just works fine! (It is equivalent to the condition `u != 0`.)

HTH, Niels

**Comment by hjmjohnson** [comment]
@ILoveGoulash Thank you for your contribution.  There are just a few final actions that need to be finalized so that we can incorporate this into the main tree.

File: Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx (hunk 1 above)
Comment (ILoveGoulash) [reply Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx]: You're right. I'll try to find the time tomorrow to fix that.

**Comment by ILoveGoulash** [comment]
Should be okay now.

**Comment by hjmjohnson** [comment]
@ILoveGoulash I rebased and removed the merge conflict, then updated the commit message for the regression test to be more descriptive.

File: Modules/Segmentation/Voronoi/test/itkVoronoiDiagram2DInfiniteLoopTest.cxx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "itkVoronoiDiagram2DGenerator.h"
+#include "itkTestingMacros.h"
+
+int
+itkVoronoiDiagram2DInfiniteLoopTest(int argc, char * argv[])
+{
+  if (argc != 1)
+  {
+    std::cerr << "Takes no parameters, but " << argc << "parameters given" << std::endl;
```
Comment (N-Dekker) [inline Modules/Segmentation/Voronoi/test/itkVoronoiDiagram2DInfiniteLoopTest.cxx:27]: When the test has no parameters, preferably make it a GoogleTest. GoogleTest unit tests are typically more "to the point", avoiding unnecessary verbosity.

File: Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx [hunk 3]
```diff
@@ -273,8 +273,10 @@ VoronoiDiagram2DGenerator<TCoordinate>::ConstructDiagram()
     buildEdges.push_back(curr);
     EdgeInfo front = curr;
     EdgeInfo back = curr;
-    while (!(rawEdges[i].empty()))
+    auto     maxStop = rawEdges[i].size();
+    while (!(rawEdges[i].empty()) && (maxStop != 0))
```
Comment (N-Dekker) [inline Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx:277]: Do I understand correctly that the check `!(rawEdges[i].empty())` is redundant now? I mean, could it just say:

```cpp
    while ( maxStop > 0 )
```

If so, I would prefer such a simplification.

Comment (N-Dekker) [reply Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx:277]: Couldn't simply be a `for` loop?

```cpp
    for ( auto maxStop = rawEdges[i].size(); maxStop != 0; --maxStop ) 
```

Maybe then rename `maxStop` to `j`:
```cpp
    for ( auto j = rawEdges[i].size(); j != 0; --j ) 
```

Comment (hjmjohnson) [reply Modules/Segmentation/Voronoi/include/itkVoronoiDiagram2DGenerator.hxx:277]: @N-Dekker   I've looked into this some more.  The problem is more difficult than the requested change you proposed. I think the issue is that we are running into a corner case of the algorithm that is not handled. 

This PR adds a test that induces an infinite loop.  The initially requested change breaks the infinite loop but changes the algorithm's behavior.  (other test starts failing).

I haven't figured out what needs to be done, but it likely requires careful consideration of the implemented algorithm and more careful identification of corner cases.

The crux of the problem is removing items from the list for processing, then adding them back to the list if none of the described cases are met (i.e. an infinite loop occurs).

**Comment by hjmjohnson** [comment]
Superseded by #6017 (merged to main) and #6052 (backported to release-5.4). Thank you @ILoveGoulash for the original report and fix — it directly informed the final implementation.
